private/tools/java/com/github/bazelbuild/rules_jvm_external/maven/MavenPublisher.java

```diff
@@ -82,101 +86,136 @@
-              + Arrays.toString(SUPPORTED_UPLOAD_SCHEMES));
+    final String repo = System.getenv("MAVEN_REPO");
+    if (Strings.isNullOrEmpty(repo)) {
+      throw new IllegalArgumentException("MAVEN_REPO environment variable must be set");
+    }
+
+    // give friendly warning if true/false not explicitly set as args[3]
+    if (!"true".equalsIgnoreCase(args[3]) && !"false".equalsIgnoreCase(args[3])) {
+      LOG.warn(
+          "Fourth argument <publish maven metadata> should be true or false. Found: {}. Defaulting"
+              + " to false.",
+          args[3]);
     }
+    boolean publishMavenMetadata = Boolean.parseBoolean(args[3]);
 
     boolean gpgSign = Boolean.parseBoolean(System.getenv("GPG_SIGN"));
     Credentials credentials =
         new BasicAuthCredentials(System.getenv("MAVEN_USER"), System.getenv("MAVEN_PASSWORD"));
     boolean useInMemoryPgpKeys = Boolean.parseBoolean(System.getenv("USE_IN_MEMORY_PGP_KEYS"));
     String signingKey = System.getenv("PGP_SIGNING_KEY");
     String signingPassword = System.getenv("PGP_SIGNING_PWD");
-    MavenSigning.SigningMetadata signingMetadata =
-        new MavenSigning.SigningMetadata(gpgSign, useInMemoryPgpKeys, signingKey, signingPassword);
+    SigningMetadata signingMetadata =
+        new SigningMetadata(gpgSign, useInMemoryPgpKeys, signingKey, signingPassword);
 
-    List<String> parts = Arrays.asList(args[0].split(":"));
-    if (parts.size() != 3) {
-      throw new IllegalArgumentException(
-          "Coordinates must be a triplet: " + Arrays.toString(parts.toArray()));
+    final ExecutorService executorService = Executors.newSingleThreadExecutor();
+
+    try {
+      run(
+          args[0],
+          args[1],
+          args[2],
+          publishMavenMetadata,
+          args.length > 4 ? args[4] : null,
+          repo,
+          credentials,
+          signingMetadata,
+          executorService);
+    } finally {
+      executorService.shutdown();
     }
+  }
 
-    Coordinates coords = new Coordinates(parts.get(0), parts.get(1), parts.get(2));
+  protected static void run(
+      String coordinates,
+      String pomPath,
+      String mainArtifactPath,
+      boolean publishMavenMetadata,
+      @Nullable String extraArtifacts,
+      String repo,
+      @Nullable Credentials credentials,
+      SigningMetadata signingMetadata,
+      Executor executor)
+      throws Exception {
 
-    // Calculate md5 and sha1 for each of the inputs
-    Path pom = Paths.get(args[1]);
-    Path mainArtifact = getPathIfSet(args[2]);
+    if (!isSchemeSupported(repo)) {
+      throw new IllegalArgumentException(
+          "Repository must be accessed via the supported schemes: "
+              + Arrays.toString(SUPPORTED_SCHEMES));
+    }
 
-    try {
-      List<CompletableFuture<Void>> futures = new ArrayList<>();
-      futures.add(upload(repo, credentials, coords, ".pom", pom, signingMetadata));
+    if (!isUploadSchemeSupported(repo)) {
```

Comment:
Since the upload schemes are a subset of the supported schemes, this means that the difference of those sets can't be used any more with the publisher. `artifactregistry` and `gs` would both be excluded.

Looking at the previous code, this same error existed, but it should be fixed.